schema_utils.py

```diff
@@ -94,6 +94,10 @@ def normalize_against_schema(
     normalized_obj = None # type: Any
 
     if schema[SCHEMA_KEY_TYPE] == SCHEMA_TYPE_BOOL:
```

Comment:
It's not OK to change the code here. The payload validator should be handling conversion of items to booleans, not the general schema validation service (since a string is simply not a boolean).

Perhaps @DubeySandeep can use enums instead of booleans in his PR, so that they are legitimately strings. Or the payload validator could convert "true" and "false" to their corresponding booleans if they are passed through the request rather than the payload.

Reply:
> It's not OK to change the code here. The payload validator should be handling conversion of items to booleans, not the general schema validation service (since a string is simply not a boolean).
> 
> Perhaps @DubeySandeep can use enums instead of booleans in his PR, so that they are legitimately strings. Or the payload validator could convert "true" and "false" to their corresponding booleans if they are passed through the request rather than the payload.

Got your point, removed the changes from schema_utils. 
We can identify an arg whether it is coming from payload/request in base.py but not in payload_validator layer, because so far our aim is to validate all the args based on request methods (PUT, etc) not based on the category, whether it is from coming payload or request. Thus handling in other way is better.
Thank you

Reply:
I'm fine with you validating it based on request methods -- wouldn't that amount to the same thing anyway? I.e. GET/DELETE are associated with request args and PUT/POST are associated with payload args?

Reply:
Yes in general it is true, but only for an exception case when we upload file via a PUT/POST.
Eg: An image uploaded via POST is associated with request args not with payloads.

Reply:
@seanlip The schema_utils.normalize functions takes inputs and normalizes it to support the given schema in case it's not possible to normalize it raises validation issue. We have done such conversion for `int` as well ([ref1],(https://github.com/oppia/oppia/blob/82e41b10482740b12c60a35fd57a7521bfc113bb/schema_utils.py#L141), [ref2](https://github.com/oppia/oppia/blob/82e41b10482740b12c60a35fd57a7521bfc113bb/schema_utils.py#L132), [ref3](https://github.com/oppia/oppia/blob/82e41b10482740b12c60a35fd57a7521bfc113bb/schema_utils.py#L153), [ref4](https://github.com/oppia/oppia/blob/82e41b10482740b12c60a35fd57a7521bfc113bb/schema_utils.py#L155)).

If we pass `'1'` to schema_utils.normalize method it returns `1`, considering this should we also handle such boolean conversion in the schema_utils? (Handling it in the payload_validator file makes it a special case)

Reply:
No. There are two reasons:

- Firstly, there is no standard string-to-boolean conversion that maps 'False' and 'false' to False. In fact, technically, the only standard one is where the empty string is False and everything else is True (i.e. the behaviour of bool(random_string)).
- Secondly, the communication between the frontend and backend is fully controlled by us. We should ensure that it is precise and does not require a normalization step that may or may not be applied. Otherwise, this is where bugs are likely to happen.